shell/server/core.js

```diff
@@ -112,53 +116,121 @@ NotificationHandle.prototype.close = function () {
   });
 };
 
-NotificationHandle.prototype.save = function (params) {
-  var self = this;
+saveFrontendRef = function (frontendRef, owner, userId) {
   return inMeteor(function () {
-    var sturdyRef = new Buffer(Random.id(20));
+    var sturdyRef = new Buffer(generateSturdyRef());
     var hashedSturdyRef = hashSturdyRef(sturdyRef);
     ApiTokens.insert({
       _id: hashedSturdyRef,
-      frontendRef: {
-        notificationHandle: self.notificationId
-      },
-      owner: params.sealFor,
+      frontendRef: frontendRef,
+      owner: owner,
+      created: new Date(),
+      userId: userId
     });
-    self.saved = true;
     return {sturdyRef: sturdyRef};
   });
 };
 
-function restoreInternal(sturdyRef, ownerPattern) {
-  // Restores `sturdyRef`, checking first that its owner matches `ownerPattern`.
+NotificationHandle.prototype.save = function (params) {
+  return saveFrontendRef({notificationHandle: this.notificationId}, params.sealFor);
+};
 
-  var hashedSturdyRef = hashSturdyRef(sturdyRef);
-  var token = ApiTokens.findOne(hashedSturdyRef);
+function checkRequirements (requirements) {
+  if (!requirements) {
+    return true;
+  }
+  for (var i in requirements) {
+    var requirement = requirements[i];
+    if (requirement.tokenValid) {
+      var token = ApiTokens.findOne({_id: requirement.tokenValid}, {fields: {permissions: 1}});
+      if (!checkRequirements(token.permissions)) {
+        return false;
+      }
+    } else if (requirement.permissionsHeld) {
+      var p = requirement.permissionsHeld;
+      var viewInfo = Grains.findOne(p.grainId, {fields: {cachedViewInfo: 1}}).cachedViewInfo;
+      var set = new PermissionSet(grainPermissions(p.grainId, p.userId, viewInfo));
```

Comment:
This will throw an exception if `viewInfo` is `null` or `undefined`. You should either pass in `{}` in such cases or change `grainPermissions()` so that it is more forgiving.